x-pack/plugins/endpoint/public/embeddables/resolver/models/indexed_process_tree.ts

```diff
@@ -15,21 +15,69 @@ import { levelOrder as baseLevelOrder } from '../lib/tree_sequencers';
 export function factory(processes: ResolverEvent[]): IndexedProcessTree {
   const idToChildren = new Map<string | undefined, ResolverEvent[]>();
   const idToValue = new Map<string, ResolverEvent>();
+  const idToAdjacent = new Map<string, AdjacentProcessMap>();
 
   for (const process of processes) {
-    idToValue.set(uniquePidForProcess(process), process);
+    const uniqueProcessPid = uniquePidForProcess(process);
+    idToValue.set(uniqueProcessPid, process);
+
     const uniqueParentPid = uniqueParentPidForProcess(process);
     const processChildren = idToChildren.get(uniqueParentPid);
+    function emptyAdjacencyMap(id: string): AdjacentProcessMap {
+      return {
+        self: id,
+        up: null,
+        down: null,
+        previous: null,
+        next: null,
+        get level(): number {
+          if (!this.up) {
+            return 1;
+          }
+          const mapAbove = idToAdjacent.get(this.up!);
+          return mapAbove ? mapAbove.level + 1 : 1;
+        },
+      };
+    }
+    const parentAdjacencyMap =
+      (uniqueParentPid && idToAdjacent.get(uniqueParentPid)) ||
+      emptyAdjacencyMap(uniqueParentPid || 'root');
+    const adjacencyMapToUpdate: AdjacentProcessMap =
+      idToAdjacent.get(uniqueProcessPid) || emptyAdjacencyMap(uniqueProcessPid);
+
     if (processChildren) {
       processChildren.push(process);
+
+      /**
+       * Update adjacency maps for current and previous entries
+       */
+      const previousProcessId = uniquePidForProcess(processChildren[processChildren.length - 2]);
```

Comment:
nit: It took me a second to figure out why we're using `-2` here, not sure if it'd be clearer, but I guess we could move this line before the `processChildren.push()` so it's more clear that it's last child 🤷‍♂

Reply:
good 👁 . I'll rearrange that.

Reply:
⚡️d it.